### PR TITLE
fix(shared): clean up zero-count entries in ConnectionLimiter to prevent memory leak

### LINKED ISSUE #3448: fix(shared): ConnectionLimiter retains zero-count map entries on Release (memory leak)
## Problem

`ConnectionLimiter` in `packages/shared/pkg/connlimit` tracks and enforces per-key concurrent connection limits. When connections are acquired and subsequently released back to zero, `Release(key)` decrements the counter to 0, but never removes the key entry from the underlying concurrent map (`smap.Map[*atomic.Int64]`).

For long-running proxy nodes (`client-proxy` and `orchestrator/pkg/tcpfirewall`) handling high request volume across ephemeral client IPs or sandboxes, inactive keys remain permanently allocated in memory with count `0`, causing steady memory accumulation over time.

## Root Cause

`Release(key)` decrements `counter.Load()` to `0` via `CompareAndSwap(current, current-1)` but lacks eviction logic when `current - 1 == 0`. Calls to `Remove(key)` are only manually issued upon sandbox deletion in orchestrator, leaving generic HTTP proxy and TCP firewall IP keys orphaned forever.

| Setting / Factor | Current Value / State | Intended / Expected |
|---|---|---|
| File / Component | `packages/shared/pkg/connlimit/limiter.go:L48-L63` | Automatic zero-count key eviction |
| Map Retention | Retains zero-count keys indefinitely ((N_{\text{total\_unique\_keys}})$) | Evicts zero-count keys on release ((N_{\text{active\_concurrent\_keys}})$) |

## Reproduction Steps

1. Acquire and release 10,000 unique keys using `TryAcquire(key, limit)` followed by `Release(key)`.
2. Inspect `limiter.connections.Count()`.
3. Observed result: `Count()` returns `10000` (10,000 zero-count map entries retained).
4. Expected result: `Count()` returns `0`.

```go
limiter := connlimit.NewConnectionLimiter()
for i := 0; i < 10000; i++ {
    key := fmt.Sprintf("key-%d", i)
    limiter.TryAcquire(key, 5)
    limiter.Release(key)
}
// limiter.connections.Count() remains 10000
```

## Technical Context

- **Files affected**: [`packages/shared/pkg/connlimit/limiter.go`](file:///home/jimmy/Desktop/DEV/e2b/infra/packages/shared/pkg/connlimit/limiter.go#L48-L63), [`packages/shared/pkg/connlimit/limiter_test.go`](file:///home/jimmy/Desktop/DEV/e2b/infra/packages/shared/pkg/connlimit/limiter_test.go#L104-L129)
- **Subsystem**: Shared / Client Proxy / Orchestrator TCP Firewall
- **Impact**: Medium (Memory accumulation on edge proxy nodes)

## Proposed Changes

| # | Change | File(s) Affected | Complexity |
|---|---|---|---|
| 1 | Atomically evict zero-count keys in `Release(key)` using `RemoveCb` | `packages/shared/pkg/connlimit/limiter.go` | Low |
| 2 | Add unit tests verifying zero-count map key eviction | `packages/shared/pkg/connlimit/limiter_test.go` | Low |

### PR DESCRIPTION
**Pull request by chill-czar**
Closes #3448

## Summary

- Automatically evict zero-count key entries from `ConnectionLimiter` map when all active connections for a key are released.
- Add atomic eviction callback `RemoveCb` in `Release(key)` within `packages/shared/pkg/connlimit/limiter.go`.
- Add unit test coverage in `packages/shared/pkg/connlimit/limiter_test.go` asserting zero-count map key cleanup and fresh re-acquisition behavior.

## Why

`ConnectionLimiter` in `packages/shared/pkg/connlimit` tracks per-key concurrent connections in an `smap.Map[*atomic.Int64]`. Previously, `Release(key)` decremented the counter to `0` via `CompareAndSwap`, but never removed the key entry from the underlying concurrent map.

For generic HTTP proxy connection limiting and client-IP rate limiting where there is no sandbox termination event to invoke `Remove(key)` manually, every historic unique key remained in memory with count `0`. On long-running edge proxy services (`client-proxy` and `orchestrator/pkg/tcpfirewall`), this resulted in steady heap memory accumulation over time.

`smap.RemoveCb` checks `v.Load() == 0` under `smap`'s internal shard lock, guaranteeing zero-count keys are safely removed without race conditions against concurrent `TryAcquire` calls.

## Test Plan

- [x] Unit tests pass: `go test -race -v ./packages/shared/pkg/connlimit/...`
- [x] Verified zero-count key eviction via `limiter.connections.Count()`. assertions
- [x] Verified concurrent acquire/release under Go race detector (`-race`)
- [x] Formatted code cleanly: `go fmt ./...`
- [x] No regressions across `packages/shared` test suite

/cc @jakubno @dobrac @ValentaTomas @arkamar @tvi

### CODE REVIEW

**Comment by chill-czar** [comment]
@codex

File: packages/shared/pkg/connlimit/limiter.go [hunk 1]
```diff
@@ -57,6 +57,12 @@ func (l *ConnectionLimiter) Release(key string) {
 			return
 		}
 		if counter.CompareAndSwap(current, current-1) {
+			if current-1 == 0 {
+				l.connections.RemoveCb(key, func(k string, v *atomic.Int64, exists bool) bool {
+					return exists && v.Load() == 0
+				})
```
[the remark this answers is not in the document]
Comment (chill-czar) [reply packages/shared/pkg/connlimit/limiter.go]: Good catch! Fixed in commit e4f15b51b using atomic tombstone state (`-1`):

1. **In `Release`**: When a counter decrements from `1` to `0`, `Release` attempts `counter.CompareAndSwap(0, -1)` to mark the counter instance as tombstoned. If `CAS(0, -1)` succeeds, `Release` calls `RemoveCb` to remove `key` from the map if it still references that tombstoned counter instance. If a concurrent `TryAcquire` acquired the counter before `CAS(0, -1)`, `CAS(0, -1)` fails and `Release` leaves the map entry intact.
2. **In `TryAcquire`**: If a counter has state `-1` (`current < 0`), `TryAcquire` detects the tombstoned counter and loops to re-fetch/insert a fresh counter instance via `getCounter`. `getCounter` atomically replaces tombstoned counters in `smap.Upsert`.
3. **Tests**: Added `TestConnectionLimiter_Concurrent/concurrent_acquire_during_release_does_not_orphan_connections` in `limiter_test.go` to verify zero untracked connections under high concurrency.

**Comment by chill-czar** [comment]
Hi @jakubno @ben-fornefeld , would appreciate a review on this when you have a moment.

**Quick summary**: Addresses a memory leak in `ConnectionLimiter` where released zero-count connection keys remained in the map. We added atomic `RemoveCb` eviction upon `Release()` to safely clean up zero-count keys without racing concurrent `TryAcquire()` calls. Unit tests and race-detector checks are passing. Thanks!

**Review by leonmeijer — COMMENTED** [review]
TryAcquire creates a counter before checking maxLimit == 0, then returns without acquiring. Because Release is never called for a rejected acquisition, each unique blocked key remains in the map at count zero, preserving an unbounded memory leak for the documented zero-limit mode. Return before getCounter when the limit is zero, or explicitly remove the unused entry on this path.

**Comment by chill-czar** [comment]
Thanks @leonmeijer for the catch!

Fixed in `8f6848eb3`:
- Added an early return at the start of `TryAcquire` when `maxLimit == 0`:
  ```go
  if maxLimit == 0 {
      return l.Count(key), false
  }
  ```
  This immediately rejects acquisition and returns `(current, false)` without calling `getCounter(key)` or inserting an unreleased map entry for new keys.
- Updated `TestConnectionLimiter_TryAcquire` in `limiter_test.go` to assert `limiter.connections.Count() == 0` for both single and multiple keys under zero-limit mode, confirming no map entries are retained.
